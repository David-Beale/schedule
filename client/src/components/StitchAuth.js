import React, {
  useEffect,
  createContext,
  useContext,
  useState,
  useMemo
} from 'react';

import {
  hasLoggedInUser,
  loginAnonymous,
  loginGoogle,
  logoutCurrentUser,
  getCurrentUser,
  addAuthenticationListener,
  removeAuthenticationListener,
  handleOAuthRedirects
} from './../stitch/authentication';

const StitchAuthContext = createContext();

export function useStitchAuth() {
  const context = useContext(StitchAuthContext);
  if (!context) {
    throw new Error(`useStitchAuth must be used within a StitchAuthProvider`);
  }
  return context;
}

export function StitchAuthProvider(props) {
  const [authState, setAuthState] = useState({
    isLoggedIn: hasLoggedInUser(),
    currentUser: getCurrentUser()
  });

  useEffect(() => {
    const authListener = {
      onUserLoggedIn: (auth, loggedInUser) => {
        if (loggedInUser) {
          setAuthState((authState) => ({
            ...authState,
            isLoggedIn: true,
            currentUser: loggedInUser
          }));
        }
      },
      // eslint-disable-next-line no-unused-vars
      onUserLoggedOut: (auth, loggedOutUser) => {
        setAuthState((authState) => ({
          ...authState,
          isLoggedIn: false,
          currentUser: null
        }));
      }
    };
    addAuthenticationListener(authListener);
    handleOAuthRedirects();
    setAuthState((state) => ({ ...state }));
    return () => {
      removeAuthenticationListener(authListener);
    };
  }, []);

  // Authentication Actions
  const handleLogin = async (provider) => {
    if (!authState.isLoggedIn) {
      switch (provider) {
        case 'anonymous':
          return loginAnonymous();
        case 'google':
          console.log('hoohle!');
          return loginGoogle();
        // eslint-disable-next-line no-empty
        default: {
        }
      }
    }
  };
  const handleLogout = async () => {
    const { isLoggedIn } = authState;
    if (isLoggedIn) {
      await logoutCurrentUser();
      setAuthState({
        ...authState,
        isLoggedIn: false,
        currentUser: null
      });
    } else {
      console.log(`can't handleLogout when no user is logged in`);
    }
  };

  // We useMemo to improve performance by eliminating some re-renders
  const authInfo = useMemo(() => {
    const { isLoggedIn, currentUser } = authState;
    const value = {
      isLoggedIn,
      currentUser,
      actions: { handleLogin, handleLogout }
    };
    return value;
  }, [authState.isLoggedIn]);
  return (
    <StitchAuthContext.Provider value={authInfo}>
      {props.children}
    </StitchAuthContext.Provider>
  );
}
