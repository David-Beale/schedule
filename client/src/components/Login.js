import React from 'react';

import { useStitchAuth } from './StitchAuth';

Login.propTypes = {};
export default function Login() {
  const { actions } = useStitchAuth();

  return (
    <div
      style={{
        height: '100vh',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center'
      }}
    >
      <div
        provider="anonymous"
        onClick={() => actions.handleLogin('anonymous')}
      >
        Log In as a Guest User
      </div>
      <div
        provider="google"
        onClick={(e) => {
          actions.handleLogin('google');
        }}
      >
        Log In with Google
      </div>
    </div>
  );
}
