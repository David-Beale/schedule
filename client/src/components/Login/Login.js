import React from 'react';
import {
  FacebookLoginButton,
  GoogleLoginButton
} from 'react-social-login-buttons';

import { useStitchAuth } from '../StitchAuth/StitchAuth';

Login.propTypes = {};
export default function Login() {
  const { actions } = useStitchAuth();

  return (
    <div
      style={{
        height: '100vh',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'column'
      }}
    >
      <FacebookLoginButton
        style={{ display: 'flex', justifyContent: 'center', width: '250px' }}
        provider="facebook"
        onClick={() => actions.handleLogin('facebook')}
      >
        Log in with Facebook
      </FacebookLoginButton>
      <GoogleLoginButton
        style={{ display: 'flex', justifyContent: 'center', width: '250px' }}
        provider="google"
        onClick={(e) => {
          actions.handleLogin('google');
        }}
      >
        Log In with Google
      </GoogleLoginButton>
    </div>
  );
}
