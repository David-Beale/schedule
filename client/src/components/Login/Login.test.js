import Login from './Login';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import React from 'react';
import { StitchAuthProvider } from '../StitchAuth/StitchAuth';

describe('Navigation', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <StitchAuthProvider>
        <Login />
      </StitchAuthProvider>,
      div
    );
  });
  it('should render properly', () => {
    const component = render(
      <StitchAuthProvider>
        <Login />
      </StitchAuthProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
