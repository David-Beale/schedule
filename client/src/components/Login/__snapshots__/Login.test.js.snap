// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="height: 100vh; display: flex; align-items: center; justify-content: center; flex-direction: column;"
      >
        <button
          style="display: flex; border: 0px; border-radius: 3px; box-shadow: rgba(0, 0, 0, 0.5) 0 1px 2px; color: rgb(255, 255, 255); cursor: pointer; font-size: 19px; margin: 5px; width: 250px; overflow: hidden; padding: 0px 10px; user-select: none; height: 50px; background: rgb(59, 89, 152); justify-content: center;"
        >
          <div
            style="align-items: center; display: flex; height: 100%;"
          >
            <div
              style="display: flex; justify-content: center; min-width: 26px;"
            >
              <svg
                height="26px"
                viewBox="0 0 90 90"
                width="26px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M90,15.001C90,7.119,82.884,0,75,0H15C7.116,0,0,7.119,0,15.001v59.998   C0,82.881,7.116,90,15.001,90H45V56H34V41h11v-5.844C45,25.077,52.568,16,61.875,16H74v15H61.875C60.548,31,59,32.611,59,35.024V41   h15v15H59v34h16c7.884,0,15-7.119,15-15.001V15.001z"
                    fill="#FFFFFF"
                  />
                </g>
              </svg>
            </div>
            <div
              style="width: 10px;"
            />
            <div
              style="text-align: left; width: 100%;"
            >
              Log in with Facebook
            </div>
          </div>
        </button>
        <button
          style="display: flex; border: 0px; border-radius: 3px; box-shadow: rgba(0, 0, 0, 0.5) 0 1px 2px; color: rgb(255, 255, 255); cursor: pointer; font-size: 19px; margin: 5px; width: 250px; overflow: hidden; padding: 0px 10px; user-select: none; height: 50px; background: rgb(203, 63, 34); justify-content: center;"
        >
          <div
            style="align-items: center; display: flex; height: 100%;"
          >
            <div
              style="display: flex; justify-content: center; min-width: 26px;"
            >
              <svg
                fill="#ffffff"
                height="26px"
                viewBox="0 0 50 50"
                width="26px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M 25.996094 48 C 13.3125 48 2.992188 37.683594 2.992188 25 C 2.992188 12.316406 13.3125 2 25.996094 2 C 31.742188 2 37.242188 4.128906 41.488281 7.996094 L 42.261719 8.703125 L 34.675781 16.289063 L 33.972656 15.6875 C 31.746094 13.78125 28.914063 12.730469 25.996094 12.730469 C 19.230469 12.730469 13.722656 18.234375 13.722656 25 C 13.722656 31.765625 19.230469 37.269531 25.996094 37.269531 C 30.875 37.269531 34.730469 34.777344 36.546875 30.53125 L 24.996094 30.53125 L 24.996094 20.175781 L 47.546875 20.207031 L 47.714844 21 C 48.890625 26.582031 47.949219 34.792969 43.183594 40.667969 C 39.238281 45.53125 33.457031 48 25.996094 48 Z "
                  />
                </g>
              </svg>
            </div>
            <div
              style="width: 10px;"
            />
            <div
              style="text-align: left; width: 100%;"
            >
              Log In with Google
            </div>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="height: 100vh; display: flex; align-items: center; justify-content: center; flex-direction: column;"
    >
      <button
        style="display: flex; border: 0px; border-radius: 3px; box-shadow: rgba(0, 0, 0, 0.5) 0 1px 2px; color: rgb(255, 255, 255); cursor: pointer; font-size: 19px; margin: 5px; width: 250px; overflow: hidden; padding: 0px 10px; user-select: none; height: 50px; background: rgb(59, 89, 152); justify-content: center;"
      >
        <div
          style="align-items: center; display: flex; height: 100%;"
        >
          <div
            style="display: flex; justify-content: center; min-width: 26px;"
          >
            <svg
              height="26px"
              viewBox="0 0 90 90"
              width="26px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M90,15.001C90,7.119,82.884,0,75,0H15C7.116,0,0,7.119,0,15.001v59.998   C0,82.881,7.116,90,15.001,90H45V56H34V41h11v-5.844C45,25.077,52.568,16,61.875,16H74v15H61.875C60.548,31,59,32.611,59,35.024V41   h15v15H59v34h16c7.884,0,15-7.119,15-15.001V15.001z"
                  fill="#FFFFFF"
                />
              </g>
            </svg>
          </div>
          <div
            style="width: 10px;"
          />
          <div
            style="text-align: left; width: 100%;"
          >
            Log in with Facebook
          </div>
        </div>
      </button>
      <button
        style="display: flex; border: 0px; border-radius: 3px; box-shadow: rgba(0, 0, 0, 0.5) 0 1px 2px; color: rgb(255, 255, 255); cursor: pointer; font-size: 19px; margin: 5px; width: 250px; overflow: hidden; padding: 0px 10px; user-select: none; height: 50px; background: rgb(203, 63, 34); justify-content: center;"
      >
        <div
          style="align-items: center; display: flex; height: 100%;"
        >
          <div
            style="display: flex; justify-content: center; min-width: 26px;"
          >
            <svg
              fill="#ffffff"
              height="26px"
              viewBox="0 0 50 50"
              width="26px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M 25.996094 48 C 13.3125 48 2.992188 37.683594 2.992188 25 C 2.992188 12.316406 13.3125 2 25.996094 2 C 31.742188 2 37.242188 4.128906 41.488281 7.996094 L 42.261719 8.703125 L 34.675781 16.289063 L 33.972656 15.6875 C 31.746094 13.78125 28.914063 12.730469 25.996094 12.730469 C 19.230469 12.730469 13.722656 18.234375 13.722656 25 C 13.722656 31.765625 19.230469 37.269531 25.996094 37.269531 C 30.875 37.269531 34.730469 34.777344 36.546875 30.53125 L 24.996094 30.53125 L 24.996094 20.175781 L 47.546875 20.207031 L 47.714844 21 C 48.890625 26.582031 47.949219 34.792969 43.183594 40.667969 C 39.238281 45.53125 33.457031 48 25.996094 48 Z "
                />
              </g>
            </svg>
          </div>
          <div
            style="width: 10px;"
          />
          <div
            style="text-align: left; width: 100%;"
          >
            Log In with Google
          </div>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
